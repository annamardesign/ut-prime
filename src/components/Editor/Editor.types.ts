import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';
import type { Schema } from 'joi';

interface Fields {
    name: string;
    title: string;
    card: string;
    editor?: {
        type: string
    }
}

interface Cards {
    id: string;
    title: string;
}

export interface Props extends React.HTMLAttributes<HTMLDivElement> {
    className?: string;
    fields: Fields[];
    cards: Cards[];
    schema?: Schema;
    onSubmit: (form: {}) => void;
}

const styles = createStyles({
    component: {
    }
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
