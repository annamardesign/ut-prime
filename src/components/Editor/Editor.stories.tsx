import React from 'react';
import { withReadme } from 'storybook-readme';
import type { Story, Meta } from '@storybook/react';

// @ts-ignore: md file and not a module
import README from './README.md';
import Editor from './index';
import type { Props } from './Editor.types';
import tree from '../test/tree';
import useToast from '../hooks/useToast';

const meta: Meta = {
    title: 'Editor',
    component: Editor,
    decorators: [withReadme(README)],
    args: {
        state: {}
    }
};
export default meta;

const Template: Story<Props> = args => {
    const {toast, submit} = useToast();
    return (
        <>
            {toast}
            <Editor
                onAdd={submit}
                onEdit={submit}
                {...args}
            />
        </>
    );
};

export const Basic: Story<Props> = Template.bind({});
Basic.args = {
    id: 1,
    object: 'tree',
    ...tree,
    layouts: {
        edit: ['edit', ['taxonomy', 'reproduction'], 'morphology', 'habitat']
    },
    onDropdown: names => Promise.resolve({
        'tree.type': [{value: 1, label: 'Conifer'}, {value: 2, label: 'Broadleaf'}],
        'tree.habitat': [
            {value: 1, label: 'Forests'},
            {value: 2, label: 'Plantations'},
            {value: 3, label: 'Riverbanks'},
            {value: 4, label: 'Rivers'},
            {value: 5, label: 'Rocky areas'},
            {value: 6, label: 'Urban'},
            {value: 7, label: 'Wetlands'}
        ]
    }),
    onGet: params => Promise.resolve({
        tree: {treeName: 'Oak', treeId: 1, treeType: 1}
    })
};

export const Loading: Story<Props> = Template.bind({});
Loading.args = {
    ...Basic.args,
    onGet: params => new Promise((resolve, reject) => {})
};

export const Design: Story<Props> = Template.bind({});
Design.args = {
    ...Basic.args,
    design: true
};

export const Tabs: Story<Props> = Template.bind({});
Tabs.args = {
    ...Basic.args,
    layouts: {
        edit: {
            orientation: 'top',
            items: [{
                icon: 'pi pi-user',
                label: 'General',
                widgets: ['edit', 'habitat']
            }, {
                label: 'Details',
                icon: 'pi pi-book',
                widgets: ['taxonomy', 'morphology']
            }, {
                icon: 'pi pi-clock',
                widgets: ['history']
            }]
        }
    }
};
