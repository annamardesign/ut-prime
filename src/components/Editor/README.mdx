# Editor
import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";

## How to use

```jsx
import Editor from 'ut-prime/core/Editor';

<Editor className='name' />;
```

<Canvas>
    <Story id="editor--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
