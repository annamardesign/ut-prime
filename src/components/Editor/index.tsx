import React from 'react';
import { Button, Card, InputText, Dropdown, InputMask, Calendar } from '../prime';

import { Styled, StyledType } from './Editor.types';
import useForm from '../hooks/useForm';
import Controller from '../Controller';
import { joiResolver } from '@hookform/resolvers/joi';

function element(field, {type = 'string', ...props} = {}) {
    const Element: React.ElementType = {
        dropdown: Dropdown,
        mask: InputMask,
        date: Calendar
    }[type] || InputText;
    return <Element {...field} {...props}/>;
}

const Editor: StyledType = ({ classes, className, fields, cards, schema, onSubmit, ...rest }) => {
    const {handleSubmit, control, formState: {errors}} = useForm({resolver: joiResolver(schema)});
    const getFormErrorMessage = (name) => {
        return errors[name] && <small className="p-error">{errors[name].message}</small>;
    };
    return (
        <div {...rest}>
            <form onSubmit={handleSubmit(onSubmit)} className="p-fluid">
                <Button type="submit" label="Submit" className="p-mt-2" />
                {(cards || []).map(({id, title}) =>
                    <Card title={title} key={id} className="p-m-2">
                        {fields.filter(({card}) => id === card).map(({name, title, editor}) =>
                            <div className="p-field p-grid" key={name}>
                                <label className="p-col-12 p-md-2">
                                    {title}
                                </label>
                                <div className="p-col-12 p-md-10">
                                    <Controller
                                        control={control}
                                        name={name}
                                        render={({field}) => element(field, editor)}
                                    />
                                </div>
                                {getFormErrorMessage(name)}
                            </div>
                        )}
                    </Card>
                )}
            </form>
        </div>
    );
};

export default Styled(Editor);
