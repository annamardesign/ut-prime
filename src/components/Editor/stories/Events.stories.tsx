import React from 'react';
import type { Story } from '@storybook/react';
import type { Props } from '../Editor.types';
import Editor from '..';
import useToast from '../../hooks/useToast';

import {input, dropdowns} from '../../test/input';

export {default} from '../Editor.stories';

const sticky = {sticky: false};

const Template: Story<Props> = ({methods, ...args}) => {
    const {toast, submit} = useToast(sticky);
    return <>
        <Editor
            id={1}
            object='input'
            onGet={() => Promise.resolve({input: {}})}
            onDropdown={() => Promise.resolve(dropdowns)}
            onAdd={submit}
            onEdit={submit}
            onFieldChange='handleFieldChange'
            methods={{
                async handleFieldChange({field, value, event}: {field: unknown, value, event: Event}) {
                    submit({field, value});
                    // throw new Error('test error');
                    // return false;
                },
                async handleAutocomplete() {
                    return [{
                        value: 1, label: 'value 1'
                    }, {
                        value: 2, label: 'value 2'
                    }, {
                        value: 3, label: 'value 3'
                    }];
                },
                async 'portal.customization.get'() {
                    return {};
                },
                ...methods
            }}
            {...args}
        />
        {toast}
    </>;
};

declare type StoryTemplate = Story<Partial<Props>> & {
    play: (context: {canvasElement: HTMLElement}) => Promise<void>
}

export const Events: StoryTemplate = Template.bind({});
Events.args = {
    ...input,
    layouts: {edit: ['left', 'center', 'right']}
};

export const EventsFormAPI: StoryTemplate = Template.bind({});
EventsFormAPI.args = {
    schema: {
        properties: {
            a: {
                type: 'integer',
                widget: {onChange: 'handleA'}
            },
            b: {
                type: 'integer',
                widget: {onChange: 'handleB'}
            },
            sum: {
                type: 'number',
                readOnly: true
            },
            discount: {
                type: 'number',
                widget: {onChange: 'handleDiscount'}
            },
            bitmask: {
                widget: {onChange: 'handleBitmask'}
            }
        }
    },
    cards: {
        edit: {
            className: 'lg:col-3',
            widgets: [
                'a',
                'b',
                'sum',
                'discount',
                'bitmask'
            ]
        }
    },
    onGet: () => Promise.resolve({a: 0, b: 0, sum: 0}),
    methods: {
        async handleA({form, value}) {
            form.setValue('sum', Number(form.getValues('b')) + Number(value));
        },
        async handleB({form, value}) {
            form.setValue('sum', Number(form.getValues('a')) + Number(value));
        },
        async handleDiscount({value, form}) {
            if (Number(value) < 0) throw new Error('Discount must be > 0'); // throwing means abort the edit and show error
            if (Number(value) >= 0) form.clearErrors('discount');
        },
        async handleBitmask({value}) {
            if (/[^01]/.test(value)) return false; // returning false means abort the edit silently
        }
    }
};
