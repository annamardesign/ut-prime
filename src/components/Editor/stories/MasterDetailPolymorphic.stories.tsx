import { userEvent, within } from '@storybook/testing-library';
import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

const common = {
    className: 'xl:col-3',
    label: 'Shape properties',
    watch: '$.selected.shape'
};

export const MasterDetailPolymorphic = () =>
    <Editor
        id={1}
        onGet={() => Promise.resolve({
            shape: [
                {name: 'Big circle', type: 'circle', radius: 100},
                {name: 'Small ellipse', type: 'ellipse', radius: 5, secondRadius: 4},
                {name: 'Big square', type: 'square', side1: 100},
                {name: 'Medium triangle', type: 'triangle', side1: 30, side2: 40, side3: 50},
                {name: 'Small rectangle', type: 'rectangle', side1: 2, side2: 4}
            ]
        })}
        onDropdown={() => Promise.resolve({})}
        schema={{
            type: 'object',
            properties: {
                shape: {
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            name: {},
                            type: {enum: ['circle', 'square', 'triangle', 'rectangle', 'ellipse']},
                            radius: {title: 'Radius', type: 'number'},
                            secondRadius: {title: 'Second radius', type: 'number'},
                            side1: {title: 'Side Length', type: 'number'},
                            side2: {title: 'Second Side Length', type: 'number'},
                            side3: {title: 'Third Side Length', type: 'number'},
                            tags: {}
                        }
                    },
                    title: '',
                    widget: {
                        type: 'table',
                        selectionMode: 'single',
                        actions: {
                            allowEdit: false
                        }
                    }
                }
            }
        }}
        cards={{
            shapes: {
                label: 'Shapes',
                className: 'xl:col-2',
                widgets: [{
                    name: 'shape',
                    widgets: ['name']
                }]
            },
            shapeCircle: {
                ...common,
                match: {type: 'circle'},
                widgets: [
                    '$.edit.shape.radius'
                ]
            },
            shapeEllipse: {
                ...common,
                match: {type: 'ellipse'},
                widgets: [
                    '$.edit.shape.radius',
                    '$.edit.shape.secondRadius'
                ]
            },
            shapeSquare: {
                ...common,
                match: {type: 'square'},
                widgets: [
                    '$.edit.shape.side1'
                ]
            },
            shapeRectangle: {
                ...common,
                match: {type: 'rectangle'},
                widgets: [
                    '$.edit.shape.side1',
                    '$.edit.shape.side2'
                ]
            },
            shapeTriangle: {
                ...common,
                match: {type: 'triangle'},
                widgets: [
                    '$.edit.shape.side1',
                    '$.edit.shape.side2',
                    '$.edit.shape.side3'
                ]
            }
        }}
        layouts={{
            edit: ['shapes', ['shapeCircle', 'shapeSquare', 'shapeTriangle', 'shapeRectangle', 'shapeEllipse']]
        }}
    />;

MasterDetailPolymorphic.play = async({canvasElement}) => {
    const canvas = within(canvasElement);
    userEvent.click(canvas.getByText('Small ellipse'));
};
