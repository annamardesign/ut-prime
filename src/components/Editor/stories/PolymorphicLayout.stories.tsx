import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

export const PolymorphicLayout = () => {
    const props = {
        typeField: 'type',
        keyField: 'id',
        onDropdown: names => Promise.resolve({}),
        schema: {
            properties: {
                name: {},
                role: {},
                registrationNumber: {}
            }
        },
        cards: {
            editPerson: {
                label: 'Person',
                className: 'col-12',
                widgets: ['name', 'role']
            },
            editOrganization: {
                label: 'Organization',
                className: 'col-12',
                widgets: ['name', 'registrationNumber']
            }
        }
    };

    return (<div className='flex flex-row'>
        <div className='flex-col flex-grow-1'>
            <Editor
                id={1}
                onGet={() => Promise.resolve({
                    id: 1,
                    type: 'person',
                    name: 'John Doe',
                    role: 'Director'
                })}
                {...props}
            />
        </div>
        <div className='flex-col flex-grow-1'>
            <Editor
                id={2}
                onGet={() => Promise.resolve({
                    id: 2,
                    type: 'organization',
                    name: 'Acme Inc.',
                    registrationNumber: '111-111-111'
                })}
                {...props}
            />
        </div>
    </div>);
};
