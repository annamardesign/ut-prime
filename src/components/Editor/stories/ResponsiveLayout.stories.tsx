import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

export const ResponsiveLayout = () =>
    <Editor
        debug
        onDropdown={names => Promise.resolve({})}
        cards={{
            a1: {label: 'A 1', className: 'sm:col-6 xl:col-4'},
            a2: {label: 'A 2'},
            b1: {label: 'B 1', className: 'sm:col-6 xl:col-4'},
            b2: {label: 'B 2'},
            b3: {label: 'B 3'},
            c: {label: 'C', className: 'md:col-6 xl:col-4'},
            d: {label: 'D', className: 'md:col-6 xl:col-4'},
            e: {label: 'E', className: 'md:col-4'},
            f1: {label: 'F 1', className: 'md:col-4'},
            f2: {label: 'F 2'},
            g: {label: 'G', className: 'md:col-4'}
        }}
        layouts={{
            edit: [
                ['a1', 'a2'],
                ['b1', 'b2', 'b3'],
                'c',
                'd',
                'e',
                ['f1', 'f2'],
                'g'
            ]
        }}
    />;
