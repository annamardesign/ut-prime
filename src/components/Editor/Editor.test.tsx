import React from 'react';

import { render } from '../test';
import { Basic } from './Editor.stories';

describe('<Editor />', () => {
    it('render equals snapshot', async() => {
        const { findByTestId } = render(<Basic />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
});
