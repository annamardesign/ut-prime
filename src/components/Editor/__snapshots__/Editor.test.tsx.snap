// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Editor /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div>
    <form
      class="p-fluid"
    >
      <button
        class="p-button p-component p-mt-2"
        type="submit"
      >
        <span
          class="p-button-label p-c"
        >
          Submit
        </span>
      </button>
    </form>
  </div>
</div>
`;
